core/src/main/java/com/linecorp/armeria/server/HttpStatusException.java

```diff
@@ -32,51 +34,69 @@
+            EXCEPTIONS[i] = new HttpStatusException(HttpStatus.valueOf(i), false, null);
         }
     }
 
     /**
      * Returns a new {@link HttpStatusException} instance with the specified HTTP status code.
      */
     public static HttpStatusException of(int statusCode) {
-        if (statusCode < 0 || statusCode >= 1000) {
-            final HttpStatus status = HttpStatus.valueOf(statusCode);
-            if (Flags.verboseExceptionSampler().isSampled(HttpStatusException.class)) {
-                return new HttpStatusException(status);
+        return of(HttpStatus.valueOf(statusCode));
+    }
+
+    /**
+     * Returns a new {@link HttpStatusException} instance with the specified HTTP status code and {@code cause}.
+     */
+    public static HttpStatusException of(int statusCode, Throwable cause) {
+        return of(HttpStatus.valueOf(statusCode), cause);
+    }
+
+    /**
+     * Returns a new {@link HttpStatusException} instance with the specified {@link HttpStatus}.
+     */
+    public static HttpStatusException of(HttpStatus status) {
+        final boolean sampled = Flags.verboseExceptionSampler().isSampled(HttpStatusException.class);
+        if (!sampled) {
+            final int statusCode = status.code();
+            if (statusCode >= 0 && statusCode < 1000) {
+                return EXCEPTIONS[statusCode];
             } else {
-                return new HttpStatusException(status, false);
+                return new HttpStatusException(HttpStatus.valueOf(statusCode), false, null);
             }
-        } else {
-            return EXCEPTIONS[statusCode];
         }
+
+        return new HttpStatusException(status, true, null);
     }
 
     /**
-     * Returns a new {@link HttpStatusException} instance with the specified {@link HttpStatus}.
+     * Returns a new {@link HttpStatusException} instance with the specified {@link HttpStatus} and
+     * {@code cause}.
      */
-    public static HttpStatusException of(HttpStatus httpStatus) {
-        requireNonNull(httpStatus, "httpStatus");
-        return of(httpStatus.code());
+    public static HttpStatusException of(HttpStatus status, Throwable cause) {
+        requireNonNull(status, "status");
+        requireNonNull(cause, "cause");
+        return new HttpStatusException(status, cause);
     }
 
     private static final long serialVersionUID = 3341744805097308847L;
 
     private final HttpStatus httpStatus;
 
     /**
-     * Creates a new instance with the specified {@link HttpStatus}.
+     * Creates a new instance.
      */
-    private HttpStatusException(HttpStatus httpStatus) {
-        super(requireNonNull(httpStatus, "httpStatus").toString());
-        this.httpStatus = httpStatus;
+    private HttpStatusException(HttpStatus httpStatus, @Nullable Throwable cause) {
+        this(httpStatus,
+             Flags.verboseExceptionSampler().isSampled(HttpStatusException.class),
```

Comment:
Having some sampler happen in factories, other in constructor is a bit mysterious. How about moving it to the factory and removing this constructor?